### PR TITLE
kernel: timer: Fix race condition in k_timer_start

### LINKED ISSUE #62908: k_timer_start asserts when called from multiple preemptive contexts
**Describe the bug**
`k_timer_start` documentation does not mention that it cannot be called from multiple preemptive contexts for the same instance. However, application may stop on an assert if one `k_timer_start` is preempted by another.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
 - What have you tried to diagnose or workaround this issue?
 - Is this a regression? If yes, have you been able to "git bisect" it to a
   specific commit?
 - ...

**To Reproduce**
1. Use modified hello_world from here: https://github.com/zephyrproject-rtos/zephyr/compare/main...nordic-krch:zephyr:timeout_assert
2. Compile it for `qemu_cortex_m3`
3. Run it for few seconds and observe following assert:
`ASSERTION FAIL [!sys_dnode_is_linked(&to->node)] @ WEST_TOPDIR/zephyr/kernel/timeout.c:110`

**Expected behavior**
Application should not fail.

**Environment (please complete the following information):**
 - Os: Linux
 - Toolchain: 0.16.1
 - Commit SHA: b0f92e13b92e5a3d54f2d3694f623aef7cf71f72

**Additional context**
It seems that the reason why assert is hit is that timer is started in a following way:
- previous timer is aborted https://github.com/zephyrproject-rtos/zephyr/blob/f74bbc8d4886c445ce12d00e11bbd363834932cd/kernel/timer.c#L157
- new values for timer are set (not under lock)
- assert checks that `timeout` is not active - not in list) https://github.com/zephyrproject-rtos/zephyr/blob/f74bbc8d4886c445ce12d00e11bbd363834932cd/kernel/timeout.c#L110
- new timeout is set (under lock)

If context is preempted after timer is aborted but before assert and in that new context same time is started then assert will be triggered.

### PR DESCRIPTION
**Pull request by sousapedro596**
The documentation suggests that k_timer_start can be invoked from ISR and preemptive contexts, however, an assertion failure occurs if one k_timer_start call preempts another for the same timer instance. This commit mitigates the issue by implementing a spinlock throughout the k_timer_start function, ensuring thread safety.

Fixes: #62908

#### Problems addressed:
* Concurrent calls to `k_timer_start` for the same timer instance from different contexts could lead to a race condition.
* The race condition could trigger an assertion failure if one `k_timer_start` call preempts another (	`__ASSERT(!sys_dnode_is_linked(&to->node), "");`
)
* Due to the race condition, the timer configuration could be incorrect, causing unexpected behavior.

Used the sample mentioned in the issue (#62908) to test and created this one and also replicated with `qemu_x86-64`: https://github.com/zephyrproject-rtos/zephyr/commit/d2d84e94b9487e0003b4835910eb2ed1022a6de0

**Note:** This is my first Pull Request to Zephyr, decided to start after checking issue #62908 (all the credit to the finding) Feedback is highly appreciated and please help me to review any potential problems with this change that I might be overlooking.

### CODE REVIEW

**Comment by peter-mitsis** [comment]
This is a nice find! However, upon further reading, I observe that we still have a mismatch between that which is documented for _k_timer_start()_ and that which is implemented.

The documentation states ...

> Attempting to start a timer that is already running is permitted. The timer's status is reset to zero and the timer begins counting down using the new duration and period values.

This would suggest that if the _duration_ parameter is set to _K_FOREVER_, then that should not result in an immediate return, but should instead reset the timer.

**Comment by p-sousa** [comment]
@peter-mitsis Thank you! Really happy to take care of that too, do you think it should be solved in the scope of this PR?

> This would suggest that if the duration parameter is set to K_FOREVER, then that should not result in an immediate return, but should instead reset the timer.

Did not tested yet, but basically, if we reschedule with `k_timer_start` with `duration` set to `K_FOREVER` the timer will keep the previous configuration. How do you think this could be addressed? 
I have a couple of concerns:
* Despite the documentation this may break existing applications that are doing this and still rely on a running timmer afterward. Not sure if I should worry about this one.
* What would be the point of rescheduling to `K_FOREVER` instead of using `k_timer_stop`? 
The only thing that comes to mind is to stop the timer without calling the `stop_fn`. But in this case, the documentation of `K_TIMER_DEFINE` and `k_timer_init`gets ambiguous too:
    > stop_fn	Function to invoke if the timer is stopped while running.
    
    Since we define a function that will be called when the timer stops, `stop_fn`, and in practice, both `k_timer_stop` and a `k_timer_start` rescheduling to `K_FOREVER` will effectively stop the timer. Would it be ok to assume that the `stop_fn` will only be called with `k_timer_Stop`?
    Moreover, if we go this way we also need to handle the case for `k_timer_start` with `duration` set to `K_FOREVER` the first time it is called (not rescheduling) and in that specific case it looks okay to behave as it is today.

At this point I may be overthinking this, can you pl share your thoughts @peter-mitsis? Do you agree with my reasoning?

**Review by cfriedt — DISMISSED** [review]

**Review by peter-mitsis — DISMISSED** [review]
As I ponder things, the issue I identified is a separate issue and does not need to be dealt with within the bounds of this PR.

File: kernel/timer.c [hunk 1]
```diff
@@ -141,7 +141,15 @@ void z_impl_k_timer_start(struct k_timer *timer, k_timeout_t duration,
 {
 	SYS_PORT_TRACING_OBJ_FUNC(k_timer, start, timer, duration, period);
 
+	/* Acquire lock to ensure during timer start configuration,
+	 * mitigating issues when k_timer_start is called concurrently
+	 * for the same timer instance from different contexts.
+	 * This addresses the race condition described in issue #62908.
```
Comment (carlescufi) [inline kernel/timer.c:147]: Please don't refer to GitHub issues in the kernel code comments. Use the commit message for that.

File: kernel/timer.c [hunk 2]
```diff
@@ -141,7 +141,15 @@ void z_impl_k_timer_start(struct k_timer *timer, k_timeout_t duration,
 {
 	SYS_PORT_TRACING_OBJ_FUNC(k_timer, start, timer, duration, period);
 
+	/* Acquire lock to ensure during timer start configuration,
+	 * mitigating issues when k_timer_start is called concurrently
```
Comment (carlescufi) [inline kernel/timer.c:145]: mitigating or eliminating? It's important to be clear here

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by sousapedro596** [comment]
@carlescufi Rephrased the whole comment and removed the mention of the issue, please re-review. 
@cfriedt @peter-mitsis please re-review.
Thank you!

**Review by peter-mitsis — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@andyross (and other Kernel collabs) could you take a look? We are pretty late in the release stages, if you want to discuss whether this should classify as release blocker this is the right moment, otherwise we could let it go for 3.6 and backport it later.

**Comment by peter-mitsis** [comment]
FWIW, my vote is to backport it at a later time as I suspect that this is a situation that we do not run into all that often.

**Review by nashif — APPROVED** [review]

File: kernel/timer.c [hunk 3]
```diff
@@ -141,7 +141,15 @@ void z_impl_k_timer_start(struct k_timer *timer, k_timeout_t duration,
 {
 	SYS_PORT_TRACING_OBJ_FUNC(k_timer, start, timer, duration, period);
 
+	/* Acquire spinlock to ensure safety during concurrent calls to
+	 * k_timer_start for scheduling or rescheduling. This is necessary
+	 * since k_timer_start can be preempted, especially for the same
+	 * timer instance.
```
Comment (andyross) [inline kernel/timer.c:147]: Not worth a -1, but if you respin I'd suggest removing this comment.  This locking strategy is 100% conventional and the comment is basically just saying "this needs a lock".  Synchronization comments should explain weird of surprising interactions, not the normal stuff.

File: kernel/timer.c [hunk 4]
```diff
@@ -141,7 +141,15 @@ void z_impl_k_timer_start(struct k_timer *timer, k_timeout_t duration,
...
+	 */
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
 	if (K_TIMEOUT_EQ(duration, K_FOREVER)) {
+		k_spin_unlock(&lock, key);
```
Comment (andyross) [inline kernel/timer.c:152]: No need to check this inside the lock, FWIW.  Saves a line of code and a tiny bit of CPU time for a pathological case.

**Review by andyross — APPROVED** [review]
Nice catch.

**Comment by andyross** [comment]
Regarding release management: this is a real fix, but seems low priority unless someone is being bitten by it in production code.  No preference on my part as to squeezing into the release vs. waiting for a backport.

**Comment by fabiobaltieri** [comment]
> Regarding release management: this is a real fix, but seems low priority unless someone is being bitten by it in production code. No preference on my part as to squeezing into the release vs. waiting for a backport.

Thanks Andy, since we are very late in the release stage, if there's no strong feeling I think it's fair to relegate this to a backport. (cc @jhedberg)

**Review by cfriedt — APPROVED** [review]
Lgtm - I would also recommend argument checking before the lock. Otherwise 👍

**Review by carlescufi — APPROVED** [review]

**Pull request merged**